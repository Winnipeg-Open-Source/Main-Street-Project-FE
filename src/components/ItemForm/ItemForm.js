import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Form from 'components/Form';
import Select from 'components/Select';
import FoodDonationForm from 'components/FoodDonationForm';
import ClothesDonationForm from 'components/ClothesDonationForm';
import HouseholdItemDonationForm from 'components/HouseholdItemDonationForm';
import MiscellaneousDonationForm from 'components/MiscellaneousDonationForm';

function CurrentForm ({ itemType, ...props }) {
    switch (itemType) {
        case 'Food':
            return <FoodDonationForm {...props} />;
        case 'Clothing':
            return <ClothesDonationForm {...props} />;
        case 'Household Item':
            return <HouseholdItemDonationForm {...props} />;
        case 'Miscellaneous':
            return <MiscellaneousDonationForm {...props} />;
        default:
            return null;
    }
}

function ItemForm ({
    className,
    itemType,
    onChange,
    onSaveClick,
    onCancelClick,
    ...props
}) {
    return (
        <Form
            className={className}
            onSaveClick={onSaveClick}
            onCancelClick={onCancelClick}
        >
            <Select
                id='item-type-select'
                name='itemType'
                value={itemType}
                onChange={onChange}
                data-testid='item-type-select'
            >
                <option>Food</option>
                <option>Clothing</option>
                <option>Household Item</option>
                <option>Miscellaneous</option>
            </Select>
            <CurrentForm itemType={itemType} onChange={onChange} {...props} />
        </Form>
    );
}

ItemForm.displayName = 'ItemForm';

ItemForm.propTypes = {
    className: PropTypes.string,
    itemType: PropTypes.string,
    locations: PropTypes.array,
    onChange: PropTypes.func,
    onSaveClick: PropTypes.func,
    onCancelClick: PropTypes.func,
};

ItemForm.defaultProps = {
    className: '',
};

export default memo(ItemForm);
