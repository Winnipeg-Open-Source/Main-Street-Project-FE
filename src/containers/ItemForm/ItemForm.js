import React, { useEffect, useState } from 'react';
import ItemFormComponent from 'components/ItemForm';
import useItems from 'hooks/context/useItems';
import useLocations from 'hooks/context/useLocations';
import useForm from 'hooks/useForm';
import useLocations from 'hooks/context/useLocations';
import useRoute from 'hooks/useRoute';
import useSaveResource from 'hooks/useSaveResource';
import { ITEMS_API_PATH } from 'constants/api';
import { ITEMS_PATH } from 'constants/paths';

const initialState = {
    itemType: 'Food',
    name: null,
    description: null,
    lineItems: [],
    notes: null,
};

function ItemForm () {
    const { onSaveItem } = useItems();
    const { state, onChange } = useForm(initialState);
    const { isLoading: isLoadingLocations, locations, onLoadLocations } = useLocations();
    const goToItems = useRoute(ITEMS_PATH);

    const { state, onChange } = useForm(initialState);
    const { lineItems, setLineItems } = useState([]);

    const addLineItem = (lineItem) => {
        setLineItems([
            ...lineItems,
            {
                ...lineItem,
                locationName: lineItem.name,
            },
        ]);
    };

    const itemQuantityChange = (itemId, lineItemId, value) => {
        setLineItems([
            ...lineItems.map(lineItem => lineItem.id === lineItemId
                ? {
                    ...lineItem,
                    quantity: value,
                }
                : lineItem
            ),
        ]);
    };

    const { handleSave } = useSaveResource(ITEMS_API_PATH, ITEMS_PATH, onSaveItem);
    const onSaveClick = () => handleSave(state);

    useEffect(() => {
        isLoadingLocations && onLoadLocations();
    }, []);

    return (
        <ItemFormComponent
            {...state}
            locations={locations}
            onChange={onChange}
            onAddLineItemClick={addLineItem}
            onItemQuantityChange={itemQuantityChange}
            onSaveClick={onSaveClick}
            onCancelClick={goToItems}
        />
    );
}

ItemForm.displayName = 'ItemForm';

export default ItemForm;
