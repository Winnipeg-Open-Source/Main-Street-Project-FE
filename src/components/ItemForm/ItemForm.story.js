import React from 'react';
import { action } from '@storybook/addon-actions';
import ItemForm from './ItemForm';
import useForm from 'hooks/useForm';

export const Default = () => {
    const { state, onChange } = useForm({ itemType: 'Food' });

    return (
        <ItemForm
            {...state}
            onChange={onChange}
            onSaveClick={action('Saved')}
            onCancelClick={action('Cancel')}
        />
    );
};

export default {
    title: 'ItemForm',
    component: ItemForm,
};
