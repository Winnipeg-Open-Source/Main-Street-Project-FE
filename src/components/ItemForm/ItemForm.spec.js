import React from 'react';
import renderWithTheme from 'tests/helpers/renderWithTheme';
import ItemForm from './ItemForm';

describe ('ItemForm', () => {
    it ('renders without crashing', () => {
        const { asFragment } = renderWithTheme(<ItemForm />);

        expect(asFragment()).toMatchSnapshot();
    });

    it ('can select food donation form', () => {
        const { queryByPlaceholderText } = renderWithTheme(<ItemForm itemType='Food' />);
        expect(queryByPlaceholderText('Retail Value ($)')).toBeInTheDocument();
    });

    it ('can select clothing donation form', () => {
        const { queryByPlaceholderText } = renderWithTheme(<ItemForm itemType='Clothing' />);
        expect(queryByPlaceholderText('Quantity')).toBeInTheDocument();
    });

    it ('can select household item donation form', () => {
        const { queryByPlaceholderText } = renderWithTheme(<ItemForm itemType='Household Item' />);
        expect(queryByPlaceholderText('Quantity')).toBeInTheDocument();
    });

    it ('can select miscellaneous donation form', () => {
        const { queryByPlaceholderText } = renderWithTheme(<ItemForm itemType='Miscellaneous' />);
        expect(queryByPlaceholderText('Quantity')).toBeInTheDocument();
    });
});
