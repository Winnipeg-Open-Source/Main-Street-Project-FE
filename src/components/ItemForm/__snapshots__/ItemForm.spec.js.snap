// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-pbJYR hNzbLI sc-pkfPC fQMEfn"
  >
    <div
      class="sc-AxjAm sc-AxiKw sc-qQxFs dNNRWZ"
    >
      <div
        class="sc-AxjAm sc-AxiKw sc-qYUWV sa-don"
      >
        <label
          class="sc-oVpqz cpWgdv className"
          color="border.light"
          font-size="10px"
          font-weight="bold"
        >
          <div
            class="sc-AxjAm sc-AxiKw iSVrPI"
            width="1"
          >
            <select
              class="sc-pZCuu kQVKTB sc-oTaid sc-pBlxj hGRJtc"
              data-testid="item-type-select"
              font-size="2,,1"
              id="item-type-select"
              name="itemType"
            >
              <option>
                Food
              </option>
              <option>
                Clothing
              </option>
              <option>
                Household Item
              </option>
              <option>
                Miscellaneous
              </option>
            </select>
            <svg
              aria-hidden="true"
              class="sc-Axmtr gXMRoh sc-fzqPZZ sc-pRFHb JnCiE"
              color="text.light"
              fill="currentcolor"
              focusable="false"
              height="24"
              role="img"
              tabindex="-1"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M7.4 8l4.6 4.6L16.6 8 18 9.4l-6 6-6-6L7.4 8z"
              />
            </svg>
          </div>
        </label>
      </div>
      <div
        class="sc-AxjAm sc-AxiKw sc-psDhf iCfenm"
      />
    </div>
  </div>
</DocumentFragment>
`;
